pkg/controller/statefulset/stateful_pod_control.go

```diff
@@ -135,15 +194,97 @@ func (spc *realStatefulPodControl) UpdateStatefulPod(set *apps.StatefulSet, pod
 	return err
 }
 
-func (spc *realStatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod *v1.Pod) error {
-	err := spc.client.CoreV1().Pods(set.Namespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})
+func (spc *StatefulPodControl) DeleteStatefulPod(set *apps.StatefulSet, pod *v1.Pod) error {
+	err := spc.objectMgr.DeletePod(pod)
 	spc.recordPodEvent("delete", set, pod, err)
 	return err
 }
 
+// ClaimsMatchDeletionPolicy returns false if the PVCs for pod are not consistent with set's PVC deletion policy.
+// An error is returned if something is not consistent. This is expected if the pod is being otherwise updated,
+// but a problem otherwise (see usage of this method in UpdateStatefulPod).
+func (spc *StatefulPodControl) ClaimsMatchDeletionPolicy(set *apps.StatefulSet, pod *v1.Pod) (bool, error) {
+	ordinal := getOrdinal(pod)
+	templates := set.Spec.VolumeClaimTemplates
+	policy := getPersistentVolumeClaimDeletePolicy(set)
+	claimShouldBeRetained := policy == apps.RetainPersistentVolumeClaimDeletePolicyType || policy == apps.DeleteOnStatefulSetDeletionOnlyPersistentVolumeClaimDeletePolicyType
+	for i := range templates {
+		claimName := getPersistentVolumeClaimName(set, &templates[i], ordinal)
+		claim, err := spc.objectMgr.GetClaim(set.Namespace, claimName)
+		switch {
+		case apierrors.IsNotFound(err):
+			if claimShouldBeRetained {
+				return false, fmt.Errorf("Expected claim %s/%s not found for %s when checking PVC deletion policy", templates[i].Namespace, claimName, pod.Name)
+			}
+		case err != nil:
+			return false, fmt.Errorf("Could not retrieve claim %s for %s when checking PVC deletion policy", claimName, pod.Name)
+		case err == nil:
+			if !claimOwnerMatchesSetAndPod(claim, set, pod) {
+				return false, nil
+			}
+		}
+	}
+	return true, nil
+}
+
+// UpdatePodClaimForDeletionPolicy updates the PVCs used by pod to match the PVC deletion policy of set.
+func (spc *StatefulPodControl) UpdatePodClaimForDeletionPolicy(set *apps.StatefulSet, pod *v1.Pod) error {
+	ordinal := getOrdinal(pod)
+	templates := set.Spec.VolumeClaimTemplates
+	for i := range templates {
+		claimName := getPersistentVolumeClaimName(set, &templates[i], ordinal)
+		claim, err := spc.objectMgr.GetClaim(set.Namespace, claimName)
+		switch {
+		case apierrors.IsNotFound(err):
+			return fmt.Errorf("Expected claim %s not found for %s when checking PVC deletion policy", claimName, pod.Name)
+		case err != nil:
+			return fmt.Errorf("Could not retrieve claim %s not found for %s when checking PVC deletion policy: %w", claimName, pod.Name, err)
+		case err == nil:
+			if !claimOwnerMatchesSetAndPod(claim, set, pod) {
+				needsUpdate := updateClaimOwnerRefForSetAndPod(claim, set, pod)
+				if needsUpdate {
+					err := spc.objectMgr.UpdateClaim(claim)
```

Comment:
How does this ensure that the OwnerReferences are correct at the API machinery side. This orphaning and adoption doesn't take into account uncached quorum reads (see DS controller https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller.go#L729). Also orphaning and adoption is typically done using Patch not Putting the whole object (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_ref_manager.go#L140:6).

Reply:
w/rt burying the feature gate, that seems necessary in order for upgrade/downgrade scenarios to work correctly. For example, if the feature gate is disabled, we do not want to stop touching ownerRefs, we may have just restarted the controller with the feature gate switched off, and so need to remove ownerRefs that were placed by the previous controller instance.

If there's a cleaner way to do that, I'd be happy to listen to any suggestions.

w/rt the empty namespaces, it seems to come from resources that are going to be created from the controller, but haven't yet been re-read from the apiserver (where presumably a missing namespace field is replaced with a "default" one?). At any rate, it seems to be necessary to do. As I mentioned to Michelle above, I think that the current statefulset controller actually does unnecessary PVC creation because of this. Adding some namespace-specific tests seems necessary and is on my list to do.

Doing a patch rather than an update makes sense, this is just my being naive. 

As for the orphaning & adoption, we only look at PVCs for pods that have already been claimed by this stateful set. So at that point we know we're looking at relevant PVCs. Probably I'm missing something here, though--can you give more details about your concern?

Reply:
I still have to evaluate if the new feature gate testing works correctly if the feature gate is enabled/disabled.

Reply:
Yeah, I think Ken is correct that UpdateClaim needs to specify prerequisites in the update operation (either RV hasn't changed, or executing a patch that preconditions the claim state prior to the mutation).

Reply:
However, right now the current behavior is not wrong (you're failing to perform the update if the object was updated underneath you, which should be a graceful retry in the controller).  We might want to mitigate error volume, but write conflicts will be rare so I don't worry too much about it.

Reply:
It sounds like this is ok for now, but for my own education, does doing a patch take care of checking for non-modifiedness or is there extra work to check a version tag or whatever?